derp/derp_server.go

```diff
@@ -1942,6 +2101,111 @@ func (s *Server) ServeDebugTraffic(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+// takeFlowsSnapshot returns a map of copies of all the flowStats for
+// this server. The output should be suitable for comparing the stats
+// for a specific flow at multiple points in time so that we can
+// calculate the bandwidth used.
+func (s *Server) takeFlowsSnapshot() map[*sclient]map[key.NodePublic]map[key.NodePublic]flowStats {
+	snap := make(map[*sclient]map[key.NodePublic]map[key.NodePublic]flowStats)
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for c := range s.flows {
+		snap[c] = make(map[key.NodePublic]map[key.NodePublic]flowStats)
+		for dstKey := range s.flows[c] {
+			snap[c][dstKey] = make(map[key.NodePublic]flowStats)
+			for srcOrFwd := range s.flows[c][dstKey] {
+				snap[c][dstKey][srcOrFwd] = *s.flows[c][dstKey][srcOrFwd]
+			}
```

Comment:
Using the [`maps.Clone`](https://pkg.go.dev/golang.org/x/exp/maps#Clone) function might simplify some of this, if you want.

Reply:
I gave it a go! Multi-level maps aren't a great match for existing libraries.